### PR TITLE
Detective Scanner Fixes

### LINKED ISSUE #18315: Detective Report no notes
It seems impossible to write on the detectives reports paper. When clicked with a pen it just opens the window

### PR DESCRIPTION
**Pull request by Davidj361**
[Changelogs]:

:cl:
fix: Detective can now write notes onto his printed scanner reports.
tweak: Detective scanner can have its logs erased via alt-click.
tweak: Detective scanner can display logs via the action button.
/:cl:

[why]:
Fixes #18315 and tweaks some functionality that is needed for the detective scanner.

### CODE REVIEW

File: code/modules/detectivework/scanner.dm [hunk 1]
```diff
@@ -165,3 +166,29 @@
 
 /proc/get_timestamp()
 	return time2text(world.time + 432000, ":ss")
+
+/obj/item/device/detective_scanner/proc/can_use(mob/user)
+	if(user && ismob(user))
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/detectivework/scanner.dm]: this should probably be `isliving`

File: code/modules/detectivework/scanner.dm [hunk 2]
```diff
@@ -165,3 +166,29 @@
...
+/obj/item/device/detective_scanner/AltClick(mob/living/user)
+	if(!can_use(user))
+		return
+	if(!LAZYLEN(log) || scanning)
+		to_chat(user, "<span class='notice'>Cannot clear logs, the scanner has no logs or is in use.</span>")
+		return
+	to_chat(user, "<span class='notice'>The scanner logs are cleared.</span>")
+	log = list()
+
+/obj/item/device/detective_scanner/verb/displayResults()
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/detectivework/scanner.dm]: Can you please use action buttons instead?

**Review by ExcessiveUseOfCobblestone — CHANGES_REQUESTED** [review]

**Comment by ExcessiveUseOfCobblestone** [comment]
Todo: Properly type `can_use` so it's a singular proc.

File: code/modules/detectivework/scanner.dm (hunk 1 above)
Comment (Cyberboss) [reply code/modules/detectivework/scanner.dm]: yes

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Cobby

File: code/modules/detectivework/scanner.dm (hunk 2 above)
Comment (Davidj361) [reply code/modules/detectivework/scanner.dm]: I prefer that action buttons are mainly combat only but if you insist.

Comment (ExcessiveUseOfCobblestone) [reply code/modules/detectivework/scanner.dm]: Oh maybe you can help pkp with the ling action buttons then :) 

On TG the default is always buttons regardless of combat usage, you should only use verbs if you have to.

Comment (Cyberboss) [reply code/modules/detectivework/scanner.dm]: Adding verbs generally isn't allowed (admin stuff being the exception, but even those use /proc shennanigans), we should write that somewhere

**Comment by ninjanomnom** [comment]
Ping for reopen if you get back to working on this

**Comment by Davidj361** [comment]
>Todo: Properly type can_use so it's a singular proc.

What do you mean?

File: code/modules/detectivework/scanner.dm [hunk 3]
```diff
@@ -165,3 +175,28 @@
 
 /proc/get_timestamp()
 	return time2text(world.time + 432000, ":ss")
+
+/obj/item/device/detective_scanner/proc/can_use(mob/user)
+	if(isliving(user))
+		if(!user.incapacitated())
+			return TRUE
```
Comment (Cyberboss) [inline code/modules/detectivework/scanner.dm]: Condense these ifs

File: code/modules/detectivework/scanner.dm [hunk 4]
```diff
@@ -16,6 +16,15 @@
...
 	var/view_check = TRUE
+	actions_types = list(/datum/action/item_action/displayDetectiveScanResults)
+
+/datum/action/item_action/displayDetectiveScanResults
+	name = "Display Forensic Scanner Results"
+
+/datum/action/item_action/displayDetectiveScanResults/Trigger()
+	var/obj/item/device/detective_scanner/scanner = target
+	if(istype(scanner))
+		scanner.displayDetectiveScanResults()
```
Comment (Cyberboss) [inline code/modules/detectivework/scanner.dm]: Pass in usr here

File: code/modules/detectivework/scanner.dm [hunk 5]
```diff
@@ -165,3 +175,28 @@
...
+	if(!can_use(user))
+		return
+	if(!LAZYLEN(log) || scanning)
+		to_chat(user, "<span class='notice'>Cannot clear logs, the scanner has no logs or is in use.</span>")
+		return
+	to_chat(user, "<span class='notice'>The scanner logs are cleared.</span>")
+	log = list()
+
+/obj/item/device/detective_scanner/proc/displayDetectiveScanResults()
+	if(!can_use(usr))
```
Comment (Cyberboss) [inline code/modules/detectivework/scanner.dm]: So you can avoid the use of usr here

File: code/modules/detectivework/scanner.dm [hunk 6]
```diff
@@ -165,3 +175,28 @@
...
+		return
+	if(!LAZYLEN(log) || scanning)
+		to_chat(usr, "<span class='notice'>The scanner has no logs or is in use.</span>")
```
Comment (Cyberboss) [inline code/modules/detectivework/scanner.dm]: here

File: code/modules/detectivework/scanner.dm [hunk 7]
```diff
@@ -165,3 +175,28 @@
...
+		return
+	to_chat(usr, "<span class='notice'><B>Scanner Report</B></span>")
+	for(var/iterLog in log)
+		to_chat(usr, iterLog)
```
Comment (Cyberboss) [inline code/modules/detectivework/scanner.dm]: and here

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/detectivework/scanner.dm [hunk 8]
```diff
@@ -165,3 +175,27 @@
...
+/obj/item/device/detective_scanner/proc/can_use(mob/user)
+	if(isliving(user) && !user.incapacitated())
+		return TRUE
+	return FALSE
+	
+/obj/item/device/detective_scanner/AltClick(mob/living/user)
+	if(!can_use(user))
+		return
+	if(!LAZYLEN(log) || scanning)
+		to_chat(user, "<span class='notice'>Cannot clear logs, the scanner has no logs or is in use.</span>")
```
Comment (Cyberboss) [inline code/modules/detectivework/scanner.dm]: There should show a different message for each reason

File: code/modules/detectivework/scanner.dm [hunk 9]
```diff
@@ -165,3 +175,27 @@
...
+		return
+	to_chat(user, "<span class='notice'>The scanner logs are cleared.</span>")
+	log = list()
+
+/obj/item/device/detective_scanner/proc/displayDetectiveScanResults(mob/living/user)
+	if(!can_use(user))
+		return
+	if(!LAZYLEN(log) || scanning)
+		to_chat(user, "<span class='notice'>The scanner has no logs or is in use.</span>")
```
Comment (Cyberboss) [inline code/modules/detectivework/scanner.dm]: ^

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by ExcessiveUseOfCobblestone** [comment]
@Davidj361 lots of objects use this "canuse" system so instead of super localized vars it would probably be better to just have a general purpose one (or better yet just put it in one of the flags).

That was a todo for me though :p

**Comment by Davidj361** [comment]
Yeah, there's no common type that uses it sadly, it's only defined for specific types sadly.

File: code/modules/detectivework/scanner.dm [hunk 10]
```diff
@@ -165,3 +175,33 @@
...
 /proc/get_timestamp()
 	return time2text(world.time + 432000, ":ss")
+
+/obj/item/device/detective_scanner/proc/can_use(mob/user)
+	if(isliving(user) && !user.incapacitated())
+		return TRUE
+	return FALSE
+	
+/obj/item/device/detective_scanner/AltClick(mob/living/user)
+	if(!can_use(user))
```
Comment (vuonojenmustaturska) [inline code/modules/detectivework/scanner.dm]: `if (!user.canUseTopic(src, be_close=TRUE))`

File: code/modules/detectivework/scanner.dm [hunk 11]
```diff
@@ -165,3 +175,33 @@
...
+		to_chat(user, "<span class='notice'>Cannot clear logs, the scanner has no logs.</span>")
+		return
+	if(scanning)
+		to_chat(user, "<span class='notice'>Cannot clear logs, the scanner is in use.</span>")
+		return
+	to_chat(user, "<span class='notice'>The scanner logs are cleared.</span>")
+	log = list()
+
+/obj/item/device/detective_scanner/proc/displayDetectiveScanResults(mob/living/user)
+	if(!can_use(user))
```
Comment (vuonojenmustaturska) [inline code/modules/detectivework/scanner.dm]: `if (!user.canUseTopic(src, be_close=TRUE))`

File: code/modules/detectivework/scanner.dm [hunk 12]
```diff
@@ -165,3 +175,33 @@
 
 /proc/get_timestamp()
 	return time2text(world.time + 432000, ":ss")
+
+/obj/item/device/detective_scanner/proc/can_use(mob/user)
```
Comment (vuonojenmustaturska) [inline code/modules/detectivework/scanner.dm]: not really needed

**Review by vuonojenmustaturska — CHANGES_REQUESTED** [review]

File: code/modules/detectivework/scanner.dm (hunk 10 above)
Comment (Davidj361) [reply code/modules/detectivework/scanner.dm]: Will do

**Pull request merged**